src/docs/binding-generics.md

```diff
@@ -236,3 +236,194 @@ The proposed API will only work with types known at compile time. In some cases
+
+If we consider the four cases:
+
+1. Non-frozen struct and at least one non-frozen generic parameter
+2. Non-frozen struct and (recursively) frozen generic parameters
+3. Frozen struct and at least one non-frozen generic parameter
+4. Frozen struct and (recursively) frozen generic parameters
+
+The cases one to three are simple from the projection perspective. The instantiated generic will be passed using indirection (as a non-frozen struct would in previous examples).
+
+The case four is problematic, as instances of a closed generic can be passed using swift lowering algorithm. Lets look into this more detail:
+
+```swift
+@frozen
+public struct Pair<T, U> {...}
+
+@frozen
+public struct FrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+public struct NonFrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+// Case 3
+public func withAtLeastOneNonFrozenParameter(data: Pair<FrozenStruct, NonFrozenStruct>) -> Pair<FrozenStruct, NonFrozenStruct> {
+    return data;
+
+// Case 4
+public func withTwoFrozenParameters(data: Pair<FrozenStruct, FrozenStruct>) -> Pair<FrozenStruct, FrozenStruct> {
+    return data;
+}
+}
+```
+
+The corresponding llvm-ir signatures of the above two functions are as follows (cases one and two would generate a signature very similar to the signature of case three).
+
+```llvm
+// Case 3
+void @"output.withAtLeastOneNonFrozenParameter(data: output.Pair<output.FrozenStruct, output.NonFrozenStruct>) -> output.Pair<output.FrozenStruct, output.NonFrozenStruct>"(ptr noalias nocapture sret(%T6output13PairVyAA0D6StructVAA0cdE0VG) %0, ptr noalias nocapture dereferenceable(32) %1)
+
+// Case 4
+{ double, double, double, double } @"output.withTwoFrozenParameters(data: output.Pair<output.FrozenStruct, output.FrozenStruct>) -> output.Pair<output.FrozenStruct, output.FrozenStruct>"(double %0, double %1, double %2, double %3)
+```
+
+Case four passes the Pair using lowering algorithm.
+
+From the user perspective we would like to have a function with the following signature:
+
+```csharp
+    public static Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair)
+```
+
+Which would then ideally call the following PInvoke.
+
+```csharp
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+    public static extern Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair);
+```
+
+This however would not work as `Pair<FrozenStruct, FrozenStruct>` is a class. Runtime would try to pass the value by reference (it does not even know the layout to begin with). We could solve this problem on the projection layer.
+
+We could generate a helper struct per instance of frozen closed generic type. This could look something like:
+
+```csharp
+[StructLayout(LayoutKind.Sequential, Size = 16)]
```

Comment:
I'm not sure whether the explicit size can be obtained at compile time for a generic instance, or it can be only done at runtime. We should be able to calculate it though in the worst scenario. 

I also wonder whether we need to even specify the explicit size at all (cc @amanasifkhalid @kotlarmilos )

Reply:
@matouskozak Do we use the size during lowering?

Reply:
Essentialy the lowering algorithm can be done just with the knowledge about the fields. 

Potentially, we could hit some issues where we assume the size is specified exactly but I don't think it would cause too much issue to work around it. I would have to try it to verify.

Reply:
> Potentially, we could hit some issues where we assume the size is specified exactly but I don't think it would cause too much problems and we should be able to work around it. I would have to try it to verify.

Do you have any example of that?

Reply:
My understanding is that we would need to specify size if the Swift type had some extra padding which is not visible on the c# side, dont know whether this is something we should worry about

Reply:
I think particularly nested struct would cause a problem because of added padding to the inner struct on C# side. For example,

```
    [StructLayout(LayoutKind.Sequential, Size = 12)]
    struct F3_S0_S0
    {
        public nint F0;
        public uint F1;

        public F3_S0_S0(nint f0, uint f1)
        {
            F0 = f0;
            F1 = f1;
        }
    }

    [StructLayout(LayoutKind.Sequential, Size = 24)]
    struct F3_S0
    {
        public sbyte F0; // Offset 0
        public F3_S0_S0 F1; // Offset 8
        public uint F2; // Offset 20

        public F3_S0(sbyte f0, F3_S0_S0 f1, uint f2)
        {
            F0 = f0;
            F1 = f1;
            F2 = f2;
        }
    }
```

The `F3_S0` should get lowered as: i8, i64, i64.

Without specifying the size of `F3_S0_S0`, the `public uint F2` would be at offset 24 (because `F3_S0_S0` size would be padded to 16), causing the current implementation of lowering algorithm to create a different lowered sequence: i8, i64, i32, i32 (because of the padding between the `F1` field from `F3_S0_S0`  and `F2` from `F3_S0`, preventing merging the two `uint` fields into single i64). 

Note: I described the process based on Mono implementation, CoreCLR might differ, but I suspect the outcome to be similar.

While we should be able to work around this by using exact sizes of individual fields, that would still require us to have knowledge about (ALL) fields that are going to be present in the struct.

Reply:
However, I forgot that this discussion is only about `frozen closed generic type` (which are known at compile time https://github.com/dotnet/runtimelab/pull/2796#discussion_r1852712009), we should be able to get the sizes ahead of time, no?

Reply:
Frozen generic struct might contain other frozen generic structs as members. 
```swift
@frozen
public struct BigStruct<T>
{
    let x : T
    let y : SmallStruct<T>
    let z : SmallStruct<Float>
}

@frozen 
public struct SmallStruct<U>
{
    public let x: U
}
```
Also closed generic type might be used to close another generic like `BigStruct<SmallStruct<Int>>`

We are able to get the size of the non-generic frozen members ahead of time, we should also have access to the layout, so we should be able to recursively calculate the size ahead of time. But I do not think we can just extract the size of the the closed / bound frozen generic from the abi.json or dylib. Do you have any ideas @stephen-hawley?

Reply:
> We are able to get the size of the non-generic frozen members ahead of time, we should also have access to the layout, so we should be able to recursively calculate the size ahead of time. But I do not think we can just extract the size of the the closed / bound frozen generic from the abi.json or dylib.

This information is available in type's metadata but we may not be able to call into it during binding time. Here are the ABI information for the example:
```json
{
"kind": "TypeDecl",
"name": "BigStruct",
"printedName": "BigStruct",
"children": [
  {
    "kind": "Var",
    "name": "x",
    "printedName": "x",
    "children": [
      {
        "kind": "TypeNominal",
        "name": "GenericTypeParam",
        "printedName": "τ_0_0"
      }
    ],
    "declKind": "Var",
    "usr": "s:14SwiftUIWrapper9BigStructV1xxvp",
    "mangledName": "$s14SwiftUIWrapper9BigStructV1xxvp",
    "moduleName": "SwiftUIWrapper",
    "isInternal": true,
    "declAttributes": [
      "HasStorage"
    ],
    "fixedbinaryorder": 0,
    "isLet": true,
    "hasStorage": true
  },
  {
    "kind": "Var",
    "name": "y",
    "printedName": "y",
    "children": [
      {
        "kind": "TypeNominal",
        "name": "SmallStruct",
        "printedName": "SwiftUIWrapper.SmallStruct<τ_0_0>",
        "children": [
          {
            "kind": "TypeNominal",
            "name": "GenericTypeParam",
            "printedName": "τ_0_0"
          }
        ],
        "usr": "s:14SwiftUIWrapper11SmallStructV"
      }
    ],
    "declKind": "Var",
    "usr": "s:14SwiftUIWrapper9BigStructV1yAA05SmallD0VyxGvp",
    "mangledName": "$s14SwiftUIWrapper9BigStructV1yAA05SmallD0VyxGvp",
    "moduleName": "SwiftUIWrapper",
    "isInternal": true,
    "declAttributes": [
      "HasStorage"
    ],
    "fixedbinaryorder": 1,
    "isLet": true,
    "hasStorage": true
  },
  {
    "kind": "Var",
    "name": "z",
    "printedName": "z",
    "children": [
      {
        "kind": "TypeNominal",
        "name": "SmallStruct",
        "printedName": "SwiftUIWrapper.SmallStruct<Swift.Float>",
        "children": [
          {
            "kind": "TypeNominal",
            "name": "Float",
            "printedName": "Swift.Float",
            "usr": "s:Sf"
          }
        ],
        "usr": "s:14SwiftUIWrapper11SmallStructV"
      }
    ],
    "declKind": "Var",
    "usr": "s:14SwiftUIWrapper9BigStructV1zAA05SmallD0VySfGvp",
    "mangledName": "$s14SwiftUIWrapper9BigStructV1zAA05SmallD0VySfGvp",
    "moduleName": "SwiftUIWrapper",
    "isInternal": true,
    "declAttributes": [
      "HasStorage"
    ],
    "fixedbinaryorder": 2,
    "isLet": true,
    "hasStorage": true
  }
],
"declKind": "Struct",
"usr": "s:14SwiftUIWrapper9BigStructV",
"mangledName": "$s14SwiftUIWrapper9BigStructV",
"moduleName": "SwiftUIWrapper",
"genericSig": "<τ_0_0>",
"sugared_genericSig": "<T>",
"declAttributes": [
  "AccessControl",
  "Frozen"
],
"conformances": [
  {
    "kind": "Conformance",
    "name": "Copyable",
    "printedName": "Copyable",
    "usr": "s:s8CopyableP",
    "mangledName": "$ss8CopyableP"
  },
  {
    "kind": "Conformance",
    "name": "Escapable",
    "printedName": "Escapable",
    "usr": "s:s9EscapableP",
    "mangledName": "$ss9EscapableP"
  }
]
},
{
"kind": "TypeDecl",
"name": "SmallStruct",
"printedName": "SmallStruct",
"children": [
  {
    "kind": "Var",
    "name": "x",
    "printedName": "x",
    "children": [
      {
        "kind": "TypeNominal",
        "name": "GenericTypeParam",
        "printedName": "τ_0_0"
      }
    ],
    "declKind": "Var",
    "usr": "s:14SwiftUIWrapper11SmallStructV1xxvp",
    "mangledName": "$s14SwiftUIWrapper11SmallStructV1xxvp",
    "moduleName": "SwiftUIWrapper",
    "declAttributes": [
      "HasStorage",
      "AccessControl"
    ],
    "fixedbinaryorder": 0,
    "isLet": true,
    "hasStorage": true,
    "accessors": [
      {
        "kind": "Accessor",
        "name": "Get",
        "printedName": "Get()",
        "children": [
          {
            "kind": "TypeNominal",
            "name": "GenericTypeParam",
            "printedName": "τ_0_0"
          }
        ],
        "declKind": "Accessor",
        "usr": "s:14SwiftUIWrapper11SmallStructV1xxvg",
        "mangledName": "$s14SwiftUIWrapper11SmallStructV1xxvg",
        "moduleName": "SwiftUIWrapper",
        "genericSig": "<τ_0_0>",
        "sugared_genericSig": "<U>",
        "implicit": true,
        "declAttributes": [
          "Transparent"
        ],
        "accessorKind": "get"
      }
    ]
  }
],
"declKind": "Struct",
"usr": "s:14SwiftUIWrapper11SmallStructV",
"mangledName": "$s14SwiftUIWrapper11SmallStructV",
"moduleName": "SwiftUIWrapper",
"genericSig": "<τ_0_0>",
"sugared_genericSig": "<U>",
"declAttributes": [
  "AccessControl",
  "Frozen"
],
"conformances": [
  {
    "kind": "Conformance",
    "name": "Copyable",
    "printedName": "Copyable",
    "usr": "s:s8CopyableP",
    "mangledName": "$ss8CopyableP"
  },
  {
    "kind": "Conformance",
    "name": "Escapable",
    "printedName": "Escapable",
    "usr": "s:s9EscapableP",
    "mangledName": "$ss9EscapableP"
  }
]
}
```

Do we have all necessary information about the bound generics?

Reply:
> Frozen generic struct might contain other frozen generic structs as members.
> 
> ```swift
> @frozen
> public struct BigStruct<T>
> {
>     let x : T
>     let y : SmallStruct<T>
>     let z : SmallStruct<Float>
> }
> 
> @frozen 
> public struct SmallStruct<U>
> {
>     public let x: U
> }
> ```
> 
> Also closed generic type might be used to close another generic like `BigStruct<SmallStruct<Int>>`
> 
> We are able to get the size of the non-generic frozen members ahead of time, we should also have access to the layout, so we should be able to recursively calculate the size ahead of time. But I do not think we can just extract the size of the the closed / bound frozen generic from the abi.json or dylib. Do you have any ideas @stephen-hawley?

I think that the way to approach this is to have a binding time type database that is as complete as possible. Given a database that has been well seeded with datatypes from libswiftCore.dylib, we should be able to calculate size/stride/alignment without having to dip into the dylib being bound.